src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -66,20 +66,120 @@ object ExpandWhens extends Pass {
     */
   type Defaults = Seq[mutable.Map[WrappedExpression, Expression]]
 
+  private def getVoid(tpe: ir.Type): WIntVoid = WIntVoid(bitWidth(tpe))
+  private def doConnect(
+    netlist:  Netlist,
+    defaults: Defaults,
+    lhs:      ir.Expression,
+    rhs:      ir.Expression,
+    info:     Info
+  ): Unit =
+    lhs match {
+      case ref: ir.RefLikeExpression =>
+        netlist(ref) = InfoExpr(info, rhs)
+      case otherLhs =>
+        // normalize nested bits expressions before check
+        simplifyBits(otherLhs) match {
+          case ir.DoPrim(PrimOps.Bits, Seq(ref: ir.RefLikeExpression), Seq(hi, lo), _) =>
+            val refWidth = bitWidth(ref.tpe)
+            val highest = refWidth - 1
+            assert(hi < refWidth && lo >= 0)
+            // if we are assigning the whole range, things are simple
+            if (hi == highest && lo == 0) {
+              netlist(ref) = InfoExpr(info, rhs)
+            } else {
+              // perform a read modify write
+              val (prevInfo, prev) = unwrap(
+                netlist.getOrElse(ref, getDefault(ref, defaults).getOrElse(WIntVoid(refWidth)))
+              )
+              val msb = if (hi == highest) { simplifyBits(rhs) }
+              else {
+                Utils.cat(simplifyBits(prev, highest, hi + 1), simplifyBits(rhs))
+              }
+              val full = if (lo == 0) { msb }
+              else {
+                Utils.cat(msb, simplifyBits(prev, lo - 1, 0))
+              }
+              netlist(ref) = InfoExpr(ir.MultiInfo(Seq(prevInfo, info)), full)
+            }
+          case other =>
+            throw new PassException(s"Invalid expression at the left hand side of an assignment: ${other.serialize}")
+        }
+    }
+  private def simplifyBits(e: ir.Expression, hi: BigInt, lo: BigInt): ir.Expression =
+    simplifyBits(Utils.bits(e, hi, lo))
+
+  /** performs special simplifications which are needed in the case of sub-word assignments to avoid false positives
+    *  in the connection check
+    */
+  private def simplifyBits(e: ir.Expression): ir.Expression = e match {
```

Comment:
I'm not sure if this code already handles this, but I recently fixed a bug in the CIRCT implementation related to SInts so I thought I'd mention it here in case. First the connector must call `asSInt` if the bit-indexed value being assigned is an `SInt`.

Then the simplifications need to look through calls to `asSInt` (i.e., having a simplification that `bits(asSInt(x), ..., ...) -> bits(x, ..., ...)`).

Reply:
I believe subword-assignment for SInt is broken in this PR right now. Thanks for pointing out how I could fix it.